### PR TITLE
feat: add propertyName operator

### LINKED ISSUE #44: add propertyname operator from jsonpath-plus
from https://jsonpath.com

given the document
```json
{
  "firstName": "John",
  "lastName" : "doe",
  "age"      : 26,
  "address"  : {
    "streetAddress": "naist street",
    "city"         : "Nara",
    "postalCode"   : "630-0192"
  },
  "phoneNumbers": [
    {
      "type"  : "iPhone",
      "number": "0123-4567-8888"
    },
    {
      "type"  : "home",
      "number": "0123-4567-8910"
    }
  ]
}
```

and the query 
```
$.firstName~
```
the output would be 
```
firstName
```

unfortunately this does not seem to be implemented in this library

this appears to come from the jsonpath-plus implementation located here: https://jsonpath-plus.github.io/JSONPath/docs/ts/

this would be useful as there is no way currently to get the yaml node containing the property name today

### PR DESCRIPTION
**Pull request by JefeDavis**
add operator to return the property name of a search
instead of the value

closes #44

Signed-off-by: Jeff Davis <jeffda@vmware.com>

### CODE REVIEW

File: pkg/yamlpath/lexer.go [hunk 1]
```diff
@@ -607,6 +644,23 @@ func lexSubPath(l *lexer) stateFn {
...
+	if l.peeked(propertyName) {
+		return lexOptionalPropertyName(l)
+	}
+	return lexOptionalArrayIndex(l)
+}
+
+func lexOptionalPropertyName(l *lexer) stateFn {
+	if l.consumed(propertyName) {
+		if l.next() != eof {
+			return l.errorf("property name operator %q may only be used at end of path", propertyName)
```
Comment (glyn) [inline pkg/yamlpath/lexer.go:657]: This error is missing a test.

Comment (glyn) [reply pkg/yamlpath/lexer.go:657]: Also the error message is inconsistent with other similar error messages in this PR since it includes `%q`. Is this intentional?

File: README.md [hunk 2]
```diff
@@ -19,12 +19,13 @@ Valid paths are strings conforming to the following BNF syntax.
...
 <child> ::= <dot child> | <bracket child>
-<dot child> ::= "." <dotted child name> | ".*"           ; named child (restricted characters) or all children
-<bracket child> ::= "[" <child names> "]"                ; named children
+<dot child> ::= "." <dotted child name> | ".*"          ; named child (restricted characters) or all children
+<bracket child> ::= "[" <child names> "]" | "[" <child names> "]~"                ; named children | property names of children
 <child names> ::= <child name> |
                   <child name> "," <child names> 
 <undotted child> ::= <dotted child name> |               ; named child (restricted characters)
                      <dotted child name><array access> | ; array access of named child
+                     <dotted child name>"~"              ; property name of child
```
Comment (glyn) [inline README.md:28]: It would be good to add a test for the case where a child name ends in `~` and make sure the behaviour is as expected. This is tricky because the set of characters allowed in a dotted child name is not documented.

File: pkg/yamlpath/path.go [hunk 3]
```diff
@@ -143,6 +145,24 @@ func newPath(l *lexer) (*Path, error) {
...
+		return propertyNameChildThen(childName, subPath), nil
+	case lexemeBracketPropertyName:
+		subPath, err := newPath(l)
+		if err != nil {
+			return nil, err
+		}
+		childNames := strings.TrimSpace(lx.val)
+		childNames = strings.TrimSuffix(childNames, propertyName)
+		childNames = strings.TrimSuffix(strings.TrimPrefix(childNames, "["), "]")
+		childNames = strings.TrimSpace(childNames)
```
Comment (glyn) [inline pkg/yamlpath/path.go:164]: This code looks like it wouldn't work if it really has to deal with child names plural.

File: pkg/yamlpath/path_test.go [hunk 4]
```diff
@@ -62,6 +62,24 @@ x:
...
+		{
+			name: "property names bracket child",
+			path: "$.store['book']~",
+			expectedStrings: []string{
+				`book
+`,
+			},
+			expectedPathErr: "",
+		},
 		{
```
Comment (glyn) [inline pkg/yamlpath/path_test.go:83]: I'd like to see a similar test with a list of property names inside the brackets.

**Review by glyn — CHANGES_REQUESTED** [review]
I've raised some questions and suggested some more tests to add. May be worth checking the effect of this PR on code coverage. Also, it would be worth running the fuzz tests for a while longer with this PR in place.

File: pkg/yamlpath/lexer.go (hunk 1 above)
Comment (JefeDavis) [reply pkg/yamlpath/lexer.go]: fixed

File: pkg/yamlpath/lexer.go [hunk 5]
```diff
@@ -534,17 +536,29 @@ func lexSubPath(l *lexer) stateFn {
...
 		childName := false
+		returnKey := false
 		for {
 			le := l.next()
 			if le == '.' || le == '[' || le == ')' || le == ' ' || le == '&' || le == '|' || le == '=' || le == '!' || le == '>' || le == '<' || le == eof {
 				l.backup()
 				break
+			} else if le == '~' {
+				if lee := l.peek(); lee != eof {
+					l.errorf("property name operator can only be used on last item in path")
```
[the remark this answers is not in the document]
Comment (JefeDavis) [reply pkg/yamlpath/lexer.go]: fixed

File: pkg/yamlpath/lexer.go [hunk 6]
```diff
@@ -572,6 +586,18 @@ func lexSubPath(l *lexer) stateFn {
...
 			return l.errorf(`missing "]" or ","`)
 		}
+		var returnKey bool
+		if l.consumed(propertyName) {
+			returnKey = true
+		}
+
+		if returnKey {
+			if le := l.peek(); le != eof {
+				l.errorf("property name operator can only be used on last item in path")
```
[the remark this answers is not in the document]
Comment (JefeDavis) [reply pkg/yamlpath/lexer.go]: fixed

File: pkg/yamlpath/lexer.go [hunk 7]
```diff
@@ -587,17 +613,28 @@ func lexSubPath(l *lexer) stateFn {
...
 		childName := false
+		returnKey := false
 		for {
 			le := l.next()
 			if le == '.' || le == '[' || le == ']' || le == ')' || le == ' ' || le == '&' || le == '|' || le == '=' || le == '!' || le == '>' || le == '<' || le == eof {
 				l.backup()
 				break
+			} else if le == '~' {
+				if lee := l.peek(); lee != eof {
+					l.errorf("property name operator can only be used on last item in path")
```
[the remark this answers is not in the document]
Comment (JefeDavis) [reply pkg/yamlpath/lexer.go]: fixed

File: README.md (hunk 2 above)
Comment (JefeDavis) [reply README.md:28]: this would work the same as if ther was a `.` in the child name, if using the doted child there is a restricted character list but you can use the bracket child syntax for this use case i.e. `$.child['childtest~']~` results in `childtest~`

File: pkg/yamlpath/path.go (hunk 3 above)
Comment (JefeDavis) [reply pkg/yamlpath/path.go:164]: This uses the same logic as the regular bracketchild syntax, it is taken care of by the bracketchildname function i added a test to verify

File: pkg/yamlpath/path_test.go (hunk 4 above)
Comment (JefeDavis) [reply pkg/yamlpath/path_test.go:83]: added

File: README.md (hunk 2 above)
Comment (JefeDavis) [reply README.md:28]: added a test for this as well

**Comment by JefeDavis** [comment]
code coverage at 95.8%

**Comment by JefeDavis** [comment]
@glyn not sure what you mean by the fuzz testing, but i believe i handled all of your comments that you left

File: pkg/yamlpath/example_test.go [hunk 8]
```diff
@@ -59,7 +59,6 @@ spec:
 
 	var buf bytes.Buffer
 	e := yaml.NewEncoder(&buf)
-	defer e.Close()
```
Comment (glyn) [inline pkg/yamlpath/example_test.go:62]: Why remove this from the example? It seems unrelated to this PR.

**Comment by glyn** [comment]
> @glyn not sure what you mean by the fuzz testing, but i believe i handled all of your comments that you left

Fuzz testing is mentioned in the project README and is described in detail [here](https://github.com/vmware-labs/yaml-jsonpath/blob/master/pkg/yamlpath/fuzz/README.md). If you run the fuzz tests for a few minutes, the fuzzer may find issues with this PR.

**Comment by glyn** [comment]
This PR changes the behaviour of the project in a backward incompatible way. Compare https://github.com/vmware-labs/yaml-jsonpath/pull/46 with https://github.com/vmware-labs/yaml-jsonpath/pull/47. Before this PR the selector `.test~` matches an object key named `test~`, but with this PR, the same selector does not match the key. This is clearly an intentional change, so perhaps the best solution would be to bump the version of the project to 2.0 to indicate the backwards incompatible change via semantic versioning.

File: pkg/yamlpath/example_test.go (hunk 8 above)
Comment (JefeDavis) [reply pkg/yamlpath/example_test.go:62]: @glyn this was causing the linter to fail due to the defer not being called if it hit the log.Fatal, it was lazy i know i can fix it up

Comment (JefeDavis) [reply pkg/yamlpath/example_test.go]: fixed

**Comment by JefeDavis** [comment]
@glyn if you are ok with it I'm going to merge this today

**Review by glyn — APPROVED** [review]
LGTM.

**Comment by JefeDavis** [comment]
@glyn I will tag off the previous version as v0.1.0 and tag this one as v0.2.0 since we haven't actually published a v1.0.0 i don't think we need to worry about the minor backwards compatibility change, it is already documented that the dotchild has a restricted character set

**Comment by glyn** [comment]
Sure. Semver allows incompatible changes in the v0.x sequence.

**Comment by ahuffman** [comment]
@glyn thanks again for working with us on this!  Good to get the original author's buy-in :).  Cheers!

**Comment by glyn** [comment]
You're welcome @ahuffman.
